File: src/compiler/parser/UnexpectedToken.java
package compiler.parser;

import compiler.lexer.token.Token;

public class UnexpectedToken extends Exception {
  public UnexpectedToken(AbstractGrammarNode top, Token token, String inputName) {
    super(formatError(top, token, inputName));
  }

  private static String formatError(AbstractGrammarNode top, Token token, String inputName) {
    final var grammarClassName = top.getClass().getSimpleName();
    final var tokenClassName = token.getClass().getSimpleName();
    final int tokenLineNumber = token.getLineNumber();
    return "\n" +
      "Unexpected token; Expected a " + grammarClassName + " but found a " + tokenClassName +
      "\n" +
      "\tat " + inputName + "(" + inputName + ":" + tokenLineNumber + ")";
  }
}

File: src/compiler/parser/UnexpectedToken.kt
package compiler.parser

import compiler.parser.Language.Token

class UnexpectedToken(top: TreeNode, token: Token, inputName: String) : Exception(formatError(top, token, inputName)) {
    companion object {
        private fun formatError(top: TreeNode, token: Token, inputName: String): String {
            val grammarClassName = top.javaClass.simpleName
            val tokenClassName = token.javaClass.simpleName
            val tokenLineNumber = token.lineInfo.number
            return "\nUnexpected token; Expected a $grammarClassName but found a $tokenClassName\n\tat $inputName($inputName:$tokenLineNumber)"
        }
    }
}
